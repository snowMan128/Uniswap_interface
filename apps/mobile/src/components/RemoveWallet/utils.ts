import { CommonActions } from '@react-navigation/core'
import { dispatchNavigationAction } from 'src/app/navigation/rootNavigation'
import { OnboardingScreens, Screens } from 'src/screens/Screens'
import { ImportType, OnboardingEntryPoint } from 'wallet/src/features/onboarding/types'

// This fast-forwards user to the same app state as if
// they have pressed "Get Started" on Landing and should now see import method view
export function navigateToOnboardingImportMethod(): void {
  dispatchNavigationAction(
    CommonActions.reset({
      index: 0,
      routes: [
        {
          name: Screens.OnboardingStack,
          state: {
            index: 1,
            routes: [
              {
                name: OnboardingScreens.Landing,
                params: {
                  entryPoint: OnboardingEntryPoint.FreshInstallOrReplace,
                  importType: ImportType.NotYetSelected,
                },
              },
              {
                name: OnboardingScreens.ImportMethod,
                params: {
                  entryPoint: OnboardingEntryPoint.FreshInstallOrReplace,
                  importType: ImportType.NotYetSelected,
                },
              },
            ],
          },
        },
      ],
    })
  )
}
