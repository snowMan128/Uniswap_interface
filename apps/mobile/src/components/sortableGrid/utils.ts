import { FlatList, ScrollView } from 'react-native'

const hasProp = <O extends object, P extends string>(
  object: O,
  prop: P
): object is O & Record<P, unknown> => {
  return prop in object
}

export const defaultKeyExtractor = <I>(item: I, index: number): string => {
  if (typeof item === 'string') {
    return item
  }

  if (typeof item === 'object' && item !== null) {
    if (hasProp(item, 'id')) {
      return String(item.id)
    }
    if (hasProp(item, 'key')) {
      return String(item.key)
    }
  }

  return String(index)
}

export const isScrollView = (scrollable: ScrollView | FlatList): scrollable is ScrollView => {
  return 'scrollTo' in scrollable
}
