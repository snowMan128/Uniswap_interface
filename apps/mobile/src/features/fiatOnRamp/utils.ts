import { FORLogo, FORQuote, FORServiceProvider } from 'wallet/src/features/fiatOnRamp/types'

export function getServiceProviderForQuote(
  quote: FORQuote | undefined,
  serviceProviders: FORServiceProvider[] | undefined
): FORServiceProvider | undefined {
  return serviceProviders?.find((sp) => sp.serviceProvider === quote?.serviceProvider)
}

export function getServiceProviderLogo(
  logos: FORLogo | undefined,
  isDarkMode: boolean
): string | undefined {
  if (!logos) {
    return
  }

  return isDarkMode ? logos.darkLogo : logos.lightLogo
}
