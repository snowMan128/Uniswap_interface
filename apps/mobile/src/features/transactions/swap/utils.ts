import {
  CurrencyField,
  TransactionState,
} from 'wallet/src/features/transactions/transactionState/types'

export function getFocusOnCurrencyFieldFromInitialState({
  focusOnCurrencyField,
  input,
  output,
  exactCurrencyField,
}: TransactionState): CurrencyField | undefined {
  if (focusOnCurrencyField) {
    return focusOnCurrencyField
  }

  if (input && exactCurrencyField === CurrencyField.INPUT) {
    return CurrencyField.INPUT
  }

  if (output && exactCurrencyField === CurrencyField.OUTPUT) {
    return CurrencyField.OUTPUT
  }

  return undefined
}
