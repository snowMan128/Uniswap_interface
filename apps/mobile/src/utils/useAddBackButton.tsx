import { NativeStackNavigationProp } from '@react-navigation/native-stack'
import React, { useEffect } from 'react'
import { OnboardingStackParamList, UnitagStackParamList } from 'src/app/navigation/types'
import { BackButton } from 'src/components/buttons/BackButton'
import { iconSizes } from 'ui/src/theme'

/**
 * Adds a back button to the navigation header if the screen is the first in the stack.
 * By default react-navigation will only show the back button if the screen is not the first one in the stack.
 */
export function useAddBackButton(
  navigation:
    | NativeStackNavigationProp<OnboardingStackParamList>
    | NativeStackNavigationProp<UnitagStackParamList>
): void {
  useEffect((): void => {
    const shouldRenderBackButton = navigation.getState().index === 0
    if (shouldRenderBackButton) {
      navigation.setOptions({
        headerLeft: () => <BackButton size={iconSizes.icon28} />,
      })
    }
  }, [navigation])
}
