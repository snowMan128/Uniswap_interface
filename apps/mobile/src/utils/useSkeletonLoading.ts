import { useEffect, useState } from 'react'

/**
 * Utility hook used to delay rendering initially so that the screen render a skeleton of placeholders
 * to allow navigation to progress before rendering heavier components that may appear as lag
 */
export function useSkeletonLoading(): boolean {
  const [enabled, setEnabled] = useState(true)

  useEffect(() => {
    setTimeout(() => setEnabled(false), 0)
  })

  return enabled
}
