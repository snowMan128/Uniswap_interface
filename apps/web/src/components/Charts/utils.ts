import { TickMarkType, UTCTimestamp } from 'lightweight-charts'

/** Compatible with ISeriesApi<'Area' | 'Candlestick'> */
export enum PriceChartType {
  LINE = 'Area',
  CANDLESTICK = 'Candlestick',
}

export enum ChartType {
  PRICE = 'Price',
  VOLUME = 'Volume',
  TVL = 'TVL', // Locked value distributed by timestamp
  LIQUIDITY = 'Liquidity', // Locked value distributed by tick
}

/**
 * Custom time formatter used to customize tick mark labels on the time scale.
 * Follows the function signature of lightweight-charts' TickMarkFormatter.
 */
export function formatTickMarks(time: UTCTimestamp, tickMarkType: TickMarkType, locale: string): string {
  const date = new Date(time.valueOf() * 1000)
  switch (tickMarkType) {
    case TickMarkType.Year:
      return date.toLocaleString(locale, { year: 'numeric' })
    case TickMarkType.Month:
      return date.toLocaleString(locale, { month: 'short', year: 'numeric' })
    case TickMarkType.DayOfMonth:
      return date.toLocaleString(locale, { month: 'short', day: 'numeric' })
    case TickMarkType.Time:
      return date.toLocaleString(locale, { hour: 'numeric', minute: 'numeric' })
    case TickMarkType.TimeWithSeconds:
      return date.toLocaleString(locale, { hour: 'numeric', minute: 'numeric', second: '2-digit' })
  }
}
