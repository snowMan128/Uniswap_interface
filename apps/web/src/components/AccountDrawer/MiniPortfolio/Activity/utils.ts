import { TransactionRequest } from '@ethersproject/abstract-provider'
import { Web3Provider } from '@ethersproject/providers'
import { t } from '@lingui/macro'
import { ChainId } from '@uniswap/sdk-core'
import { DutchOrder } from '@uniswap/uniswapx-sdk'
import { getYear, isSameDay, isSameMonth, isSameWeek, isSameYear } from 'date-fns'
import { TransactionStatus } from 'graphql/data/__generated__/types-and-hooks'
import PERMIT2_ABI from 'uniswap/src/abis/permit2.json'
import { Permit2 } from 'uniswap/src/abis/types'
import { didUserReject } from 'utils/swapErrorToUserReadableMessage'

import { PERMIT2_ADDRESS } from '@uniswap/permit2-sdk'
import { BigNumber, ContractTransaction } from 'ethers/lib/ethers'
import { useContract } from 'hooks/useContract'
import { useCallback } from 'react'
import { SignatureType } from 'state/signatures/types'
import { useAsyncData } from 'utilities/src/react/hooks'
import { Activity } from './types'

interface ActivityGroup {
  title: string
  transactions: Array<Activity>
}

const sortActivities = (a: Activity, b: Activity) => b.timestamp - a.timestamp

export const createGroups = (activities: Array<Activity> = [], hideSpam = false) => {
  if (activities.length === 0) return []
  const now = Date.now()

  const pending: Array<Activity> = []
  const today: Array<Activity> = []
  const currentWeek: Array<Activity> = []
  const last30Days: Array<Activity> = []
  const currentYear: Array<Activity> = []
  const yearMap: { [key: string]: Array<Activity> } = {}

  // TODO(cartcrom): create different time bucket system for activities to fall in based on design wants
  activities.forEach((activity) => {
    if (hideSpam && activity.isSpam) {
      return
    }

    const addedTime = activity.timestamp * 1000
    if (activity.status === TransactionStatus.Pending) {
      switch (activity.offchainOrderDetails?.type) {
        case SignatureType.SIGN_LIMIT:
          // limit orders are only displayed in their own pane
          break
        default:
          pending.push(activity)
      }
    } else if (isSameDay(now, addedTime)) {
      today.push(activity)
    } else if (isSameWeek(addedTime, now)) {
      currentWeek.push(activity)
    } else if (isSameMonth(addedTime, now)) {
      last30Days.push(activity)
    } else if (isSameYear(addedTime, now)) {
      currentYear.push(activity)
    } else {
      const year = getYear(addedTime)

      if (!yearMap[year]) {
        yearMap[year] = [activity]
      } else {
        yearMap[year].push(activity)
      }
    }
  })
  const sortedYears = Object.keys(yearMap)
    .sort((a, b) => parseInt(b) - parseInt(a))
    .map((year) => ({ title: year, transactions: yearMap[year] }))

  const transactionGroups: Array<ActivityGroup> = [
    { title: t`Pending`, transactions: pending.sort(sortActivities) },
    { title: t`Today`, transactions: today.sort(sortActivities) },
    { title: t`This week`, transactions: currentWeek.sort(sortActivities) },
    { title: t`This month`, transactions: last30Days.sort(sortActivities) },
    { title: t`This year`, transactions: currentYear.sort(sortActivities) },
    ...sortedYears,
  ]

  return transactionGroups.filter(({ transactions }) => transactions.length > 0)
}

// TODO(WEB-3594): just use the uniswapx-sdk when getCancelMultipleParams is available

interface SplitNonce {
  word: BigNumber
  bitPos: BigNumber
}

function splitNonce(nonce: BigNumber): SplitNonce {
  const word = nonce.div(256)
  const bitPos = nonce.mod(256)
  return { word, bitPos }
}

// Get parameters to cancel multiple nonces
// source: https://github.com/Uniswap/uniswapx-sdk/pull/112
function getCancelMultipleParams(noncesToCancel: BigNumber[]): {
  word: BigNumber
  mask: BigNumber
}[] {
  const splitNonces = noncesToCancel.map(splitNonce)
  const splitNoncesByWord: { [word: string]: SplitNonce[] } = {}
  splitNonces.forEach((splitNonce) => {
    const word = splitNonce.word.toString()
    if (!splitNoncesByWord[word]) {
      splitNoncesByWord[word] = []
    }
    splitNoncesByWord[word].push(splitNonce)
  })
  return Object.entries(splitNoncesByWord).map(([word, splitNonce]) => {
    let mask = BigNumber.from(0)
    splitNonce.forEach((splitNonce) => {
      mask = mask.or(BigNumber.from(2).pow(splitNonce.bitPos))
    })
    return { word: BigNumber.from(word), mask }
  })
}

function getCancelMultipleUniswapXOrdersParams(encodedOrders: string[], chainId: ChainId) {
  const nonces = encodedOrders
    .map((encodedOrder) => DutchOrder.parse(encodedOrder, chainId))
    .map((order) => order.info.nonce)
  return getCancelMultipleParams(nonces)
}

export async function cancelMultipleUniswapXOrders({
  encodedOrders,
  chainId,
  permit2,
  provider,
}: {
  encodedOrders: string[]
  chainId: ChainId
  permit2: Permit2 | null
  provider?: Web3Provider
}) {
  const cancelParams = getCancelMultipleUniswapXOrdersParams(encodedOrders, chainId)
  if (!permit2 || !provider) return
  try {
    const transactions: ContractTransaction[] = []
    for (const params of cancelParams) {
      const tx = await permit2.invalidateUnorderedNonces(params.word, params.mask)
      transactions.push(tx)
    }
    return transactions
  } catch (error) {
    if (!didUserReject(error)) console.error(error)
    return undefined
  }
}

async function getCancelMultipleUniswapXOrdersTransaction(
  encodedOrders: string[],
  chainId: ChainId,
  permit2: Permit2
): Promise<TransactionRequest | undefined> {
  const cancelParams = getCancelMultipleUniswapXOrdersParams(encodedOrders, chainId)
  if (!permit2 || cancelParams.length === 0) return
  try {
    const tx = await permit2.populateTransaction.invalidateUnorderedNonces(cancelParams[0].word, cancelParams[0].mask)
    return {
      ...tx,
      chainId,
    }
  } catch (error) {
    console.error('could not populate cancel transaction')
    return undefined
  }
}

export function useCreateCancelTransactionRequest(
  params:
    | {
        encodedOrders?: string[]
        chainId: ChainId
      }
    | undefined
): TransactionRequest | undefined {
  const permit2 = useContract<Permit2>(PERMIT2_ADDRESS, PERMIT2_ABI, true)
  const transactionFetcher = useCallback(() => {
    if (!params || !params.encodedOrders || params.encodedOrders.filter(Boolean).length === 0 || !permit2) return
    return getCancelMultipleUniswapXOrdersTransaction(params.encodedOrders, params.chainId, permit2)
  }, [params, permit2])

  return useAsyncData(transactionFetcher).data
}
