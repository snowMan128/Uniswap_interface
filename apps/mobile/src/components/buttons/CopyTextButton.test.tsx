import { act, fireEvent, render } from 'src/test/test-utils'
import { setClipboard } from 'wallet/src/utils/clipboard'
import { CopyTextButton } from './CopyTextButton'

jest.mock('wallet/src/utils/clipboard')

describe(CopyTextButton, () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.useRealTimers()
  })

  it('renders without error', () => {
    const tree = render(<CopyTextButton copyText="copy text" />)

    expect(tree).toMatchSnapshot()
  })

  it('copies text when pressed', async () => {
    const { getByText } = render(<CopyTextButton copyText="copy text" />)

    const button = getByText('Copy')
    await act(async () => {
      fireEvent.press(button)
    })

    expect(setClipboard).toHaveBeenCalledWith('copy text')
  })

  it('changes button text when text is copied and brings back original text after timeout', async () => {
    const { queryByText, getByText } = render(<CopyTextButton copyText="copy text" />)

    const button = getByText('Copy')
    await act(async () => {
      fireEvent.press(button)
    })

    expect(queryByText('Copied')).toBeTruthy()

    await act(async () => {
      jest.advanceTimersByTime(2000)
    })

    expect(queryByText('Copy')).toBeTruthy()
  })
})
