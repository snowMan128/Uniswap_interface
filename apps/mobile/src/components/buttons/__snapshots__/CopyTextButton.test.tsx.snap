// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyTextButton renders without error 1`] = `
<View
  cancelable={true}
  disabled={false}
  focusable={true}
  minPressDuration={0}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#2222220D",
      "borderBottomColor": "transparent",
      "borderBottomLeftRadius": 16,
      "borderBottomRightRadius": 16,
      "borderBottomWidth": 1,
      "borderLeftColor": "transparent",
      "borderLeftWidth": 1,
      "borderRightColor": "transparent",
      "borderRightWidth": 1,
      "borderStyle": "solid",
      "borderTopColor": "transparent",
      "borderTopLeftRadius": 16,
      "borderTopRightRadius": 16,
      "borderTopWidth": 1,
      "flexDirection": "row",
      "gap": 8,
      "height": "auto",
      "justifyContent": "center",
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 12,
      "paddingRight": 12,
      "paddingTop": 12,
    }
  }
>
  <svg
    color="#222222"
    data-file-name="SvgCopySheets"
    height={20}
    size={20}
    width={20}
  />
  <Text
    maxFontSizeMultiplier={1.2}
    style={
      {
        "color": "#222222",
        "fontFamily": "Basel-Medium",
        "fontSize": 19,
        "fontWeight": "500",
        "lineHeight": 24,
      }
    }
    suppressHighlighting={true}
  >
    Copy
  </Text>
</View>
`;
